### PR TITLE
Enable all pkg/ for QEMU testing

### PR DESCRIPTION
**Pull request by hugelgupf**
- Puts testdata in VM with each test
- Allows naming conflicts between tests (tests are now in their fully-qualified directory path)
- Test results are emitted in the VM via JSON (go test -json, go tool test2json output) and read by test framework instead of TAP protocol

All this just so I can write a nice clean gpio mock test :)

### CODE REVIEW

**Comment by hugelgupf** [comment]
cc @JulienVdG I intend to push some of this back upstream to tastevin

File: integration/gotest_test.go [hunk 1]
```diff
@@ -100,59 +153,55 @@ func TestGoTest(t *testing.T) {
...
+			JSONLessTestLineWriter(t, "serial"),
+		),
 	})
 	defer cleanup()
 
-	// Tests are run and checked in sorted order.
-	bases := []string{}
-	for _, test := range tests {
-		bases = append(bases, path.Base(test))
+	if err := q.ExpectTimeout("GoTest Done", 60*time.Second); err != nil {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply integration/gotest_test.go]: done

File: integration/gotest_test.go [hunk 2]
```diff
@@ -25,11 +28,8 @@ func testPkgs(t *testing.T) []string {
 	cmd := exec.Command("go", "list",
 		"github.com/u-root/u-root/cmds/core/...",
 		"github.com/u-root/u-root/cmds/boot/...",
-		// TODO: only running tests in cmds because tests in pkg have
-		// duplicate names which confuses the test runner. This should
-		// get fixed.
-		// "github.com/u-root/u-root/xcmds/...",
-		// "github.com/u-root/u-root/pkg/...",
+		"github.com/u-root/u-root/pkg/...",
+		"github.com/u-root/u-root/cmds/exp/...",
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply integration/gotest_test.go]: done

File: pkg/uio/multiwriter.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"io"
+)
+
+type multiCloser struct {
+	io.Writer
+	writers []io.Writer
+}
+
+func (mc *multiCloser) Close() error {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uio/multiwriter.go]: done

File: pkg/uio/multiwriter.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		if c, ok := w.(io.Closer); ok {
+			if err := c.Close(); err != nil {
+				allErr = err
+			}
+		}
+	}
+	return allErr
+}
+
+func ClosingMultiWriter(w ...io.Writer) io.WriteCloser {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uio/multiwriter.go]: done

File: pkg/uio/linewriter.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+// license that can be found in the LICENSE file.
+
+package uio
+
+import (
+	"bytes"
+	"io"
+)
+
+type LineWriter interface {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uio/linewriter.go]: done

File: pkg/uio/linewriter.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		if len(buf) != 0 {
+			fsw.w.OneLine(buf)
+		}
+	}
+	fsw.buffer = nil
+}
+
+// Write implements io.Writer and buffers p until at least one full line is
+// received.
+func (fsw *fullLineWriter) Write(p []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uio/linewriter.go]: I'm not sure what to do here yet, but the transformer seems like overkill to me

**Comment by hugelgupf** [comment]
PTAL.

In particular, take a look at this VM test: https://github.com/u-root/u-root/pull/1296/commits/89cf201b1a917b1c452b6a59a50546095714c2dd

I'm working on making that a little more useful.

**Comment by bjakobson** [comment]
It seemed to fail because of the length of the output. It was too large to display to browser.

**Comment by hugelgupf** [comment]
> It seemed to fail because of the length of the output. It was too large to display to browser.

Yeah -- moving some packages around in other PRs has caused my package blacklist here to be out of date. Fix it in a minute.

**Comment by bjakobson** [comment]
Have you downloaded or imported "github.com/u-root/u-root/integration/testcmd/uinit-734992617/uinit/.bb" into the right directory. The error message is saying it cannot find it within your designated path. Does it seem to work fine for you? I am seeing an error when you merged it.

**Pull request merged**